### ISSUE TITLE
Metrics API exposure

### ISSUE THREAD

**Issue by aluzzardi** [docker/docker#31107]
#25820 added metrics to the engine (great).

At that point it was decided to expose metrics on an entirely different endpoint, controlled via `--metrics-addr` (an experimental flag).

This raises a few concerns regarding exposing an unprotected port. For some exposing a plain text port is fine, for others leaking container IDs etc is bad.

There are a few ways to get around that, one would be to use the TLS configuration but it feels pretty meh to me.

What I'd suggest is to expose the `/metrics` endpoint on the regular Docker API. One can easily achieve the `--metrics-addr` behavior by running a container on the machine that listens to a port and exposes just the `/metrics` endpoint of the API (by querying the local daemon).

One could also chose to deploy this same service with an overlay network, thus not requiring to bind the external port.

Long story short: I believe exposing metrics under the same API server leads to a more consistent exposure level (TLS, authentication, ...) while enabling the same use case as `--metrics-addr` and then some.

Thoughts?

/cc @crosbymichael @stevvooe @dhiltgen @vieux @icecrime

**Comment by stevvooe** [comment]
@aluzzardi The `/metrics` endpoint is a read only API. Exporting the docker API is not granular enough, in practice, to allow prometheus to only hit the metrics endpoint while avoiding the exposure. By providing different paths for these API, operators can export the `/metrics` API with confidence that they won't be giving the keys away to the castle. Valid configurations that only bind the docker unix socket can export the `/metrics` endpoint with confidence, ensuring that operational staff don't have to make compromises in their infrastructure.

This was a very carefully thought through decision and I don't think we should reconsider it. This differing consistency level is very intentional. If we need to bring TLS to the `/metrics` endpoint, then we can, but let's not require exposing the whole API.

**Comment by aluzzardi** [comment]
Paging security @diogomonica @cyli

**Comment by stevvooe** [comment]
@aluzzardi I think the right solution is to retain the `--metrics-addr` flag but also bind `/metrics` to the "docker API" by default (not actually part of the API but bound to the same HTTP server). We should also include some configuration for authentication of `--metrics-addr` endpoint that will be compatible with auth supported by prometheus scrapers.

cc @juliusv @brian-brazil

**Comment by juliusv** [comment]
Not exposing the entire Docker API to Prometheus could be better for the general case, yeah. Just put comparable TLS+auth (but different client-side credentials than the API port) on the metrics port? Prometheus supports all the common HTTP auth schemes (basic auth, bearer token, and client certs).

**Comment by cyli** [comment]
Docker currently does not have any default AuthZ on the API, so giving access to the docker API is equivalent to giving admin access to that particular node, which seems not great to give to just a monitoring service.

I believe there have been discussions about a basic default read/write AuthZ plugin in docker (cc @NathanMcCauley @alexmavr) - which seems like a better way to expose the `/metrics` API.  However, that doesn't exist yet, and in the meantime, binding a separate port is one way of providing the needed Authz, and agreed providing authentication would be good.

> What I'd suggest is to expose the /metrics endpoint on the regular Docker API. One can easily achieve the --metrics-addr behavior by running a container on the machine that listens to a port and exposes just the /metrics endpoint of the API (by querying the local daemon).

> One could also chose to deploy this same service with an overlay network, thus not requiring to bind the external port.

If an admin has do to a bunch of extra work (setting up an authenticated proxy to the docker API) in order to use the metrics API, they're probably not going to bother and just give Prometheus the credentials to the entire Docker API, which I don't think we should encourage.  Possibly if we could provide and maintain such a service as an installable plugin, with configurable auth, though.

**Comment by diogomonica** [comment]
Separate port w/ MTLS by default seems like the winning move.

**Comment by aluzzardi** [comment]
@diogomonica `MTLS` would be Docker's `--tls-*` flags?

**Comment by diogomonica** [comment]
Well, after thinking a bit, adding a port seems like a copout for the lack of AuthN/Z.

**Comment by stevvooe** [comment]
> Well, after thinking a bit, adding a port seems like a copout for the lack of AuthN/Z.

Not really. Separating metrics and main API is a time honored technique. Let's not make this more complicated than it is.

We already have a good compromise:

1. Add `/metrics` to standard API.
2. Preserve `--metrics-addr` flag to allow operators to have control over binding. Include ability to configure tls client and server certs.

Both of these will support secure configuration in a variety of production environments. __We also avoid the issue of exposing the entire runtime API to a bound port just to expose the metrics.__

**Comment by diogomonica** [comment]
Ultimately I think we will still want to authorize what specific metrics a remote endpoint will have access to, so we're just punting this problem for later.

Why don't we:
1 - Add `/metrics` to standard API
2 - Provide people with a one-line container/service that only exposes `/metrics` on another port?

**Comment by stevvooe** [comment]
> 2 - Provide people with a one-line container/service that only exposes /metrics on another port?

If we make this too hard, people will opt for exposing the entire docker API, rather than just leaving it on a unix socket.

The ideal situation here is to have docker API on the unix socket with an exposed metrics API on a port. Requiring a "one-liner" and separate service for it, when we could just build it in, will not regress deployments to the ideal case.

**Comment by diogomonica** [comment]
> If we make this too hard, people will opt for exposing the entire docker API, rather than just leaving it on a unix socket.

Agreed, hence why it has to be trivial.

I genuinely believe that solving this problem for `/*insert-api-here*` is of more value to the security of our users than solving the specific case for `/metrics`

**Comment by stevvooe** [comment]
> I genuinely believe that solving this problem for /*insert-api-here* is of more value to the security of our users than solving the specific case for /metrics

Yes, but when is that solution coming along?

I hate saying this one, but perfect is the enemy of the good. Supporting `--metrics-addr`, `--metrics-cert` and `--metrics-client-cert` is a pragmatic harm minimization approach that will achieve the a solid result.

**Comment by aluzzardi** [comment]
If `/metrics` were exposed through the Docker API, exposing an additional port on every machine with **just** the metrics (and nothing else) would be trivial to do on **top** (e.g. as a service) rather than baked into the engine.

Just tried the experiment myself and it works. For the purpose of the test I've exposed `/containers/json` rather than `/metrics` because it currently doesn't exist in the API.

`docker-compose.yml`:
```yml
version: '3'

services:
    metrics-exposer:
        image: aluzzardi/nginx-metrics-exposer
        ports:
            # FIXME: Those should be local bound.
            - 8080:80
        volumes:
            - /var/run/docker.sock:/tmp/docker.sock
        deploy:
            mode: global
```

`aluzzardi/nginx-metrics-exposer` was built with:
```Dockerfile
FROM nginx:1.11.10-alpine
COPY nginx.conf /etc/nginx/nginx.conf
CMD ["nginx"]
```

and `nginx.conf`:
```
daemon off;
user root;
worker_processes  1;

events {
	    worker_connections  1024;
}

http {
	upstream docker {
		server unix://tmp/docker.sock;
	}

	server {
		listen       80;
		server_name  localhost;

		location /metrics {
			proxy_pass   http://docker/containers/json;
		}
	}
}
```

And that's it.

I don't see why we have to add that to the engine. I could use TLS in my nginx if I wanted to. I could decide to expose `/metrics` only on an internal overlay rather than a host bound port.

[Configuration here https://github.com/aluzzardi/docker-metrics-exposer]

[PS: @dnephin @mavenugo is there a way to bind a port locally in compose?]

**Comment by stevvooe** [comment]
@aluzzardi Fair enough, but I think a lot of users will just bind the docker API directly to the main port unless we make this clear. This also costs an extra process and service to monitor.

**Comment by diogomonica** [comment]
@stevvooe we totally agree with you. This has to be absolutely trivial/well documented/obvious.

**Comment by cpuguy83** [comment]
BTW, looking at a metrics collector plugin that will just get it's own metrics socket (or shared among all plugins, rather).

**Comment by aluzzardi** [comment]
Reviving this. We need to move metrics out of experimental soon-ish - metrics address or engine API?

/cc @diogomonica

**Comment by aluzzardi** [comment]
How about this:

We add metrics to the engine API (because, no matter what, it makes sense), and we keep --metrics-addr experimental. Then we'll decide whether to keep it or not.

Sounds good?

**Comment by cpuguy83** [comment]
I don't see a good reason to keep `--metrics-addr` experimental if the metrics themselves are not.

**Comment by stevvooe** [comment]
@aluzzardi How about we just do both?

**Comment by olljanat** [comment]
@aluzzardi what is plan with this one? Are you adding metrics to standard API and/or marking metrics-addr as non-experimental?

**Comment by cpuguy83** [comment]
Metrics are available through the plugin interface now.

**Comment by olljanat** [comment]
@cpuguy83 so you mean that example https://docs.docker.com/config/thirdparty/prometheus/ should be updated to say that this is deprecated method and https://prometheus.io/docs/guides/cadvisor/ should be used instead of it?

**Comment by cpuguy83** [comment]
No. I mean that the /metrics endpoint is be exposed to the "MetricsCollector" plugin type even when not experimental.

**Comment by olljanat** [comment]
Aha. You mean the one which was implemented on #32874 but is the issue then that no one have not implemented MetricsCollector plugin for Prometheus and that why documentation is still proposing to use ```metrics-addr```?

Other issue which I can see with that one is that it is not supported by Windows like experimental solution based on documentation is.

**Comment by J0WI** [comment]
Meanwhile TLS auth becomes a common feature for Prometheus exporters: https://pkg.go.dev/github.com/prometheus/exporter-toolkit/https?readme=expanded#section-readme
